packages/flutter/lib/src/widgets/framework.dart

```diff
@@ -1108,49 +1108,74 @@ abstract class Element implements BuildContext {
     assert(_debugLifecycleState == _ElementLifecycle.inactive);
     _parent = parent;
     _updateDepth(_parent.depth);
-    _activateRecursively(this);
+    _activateRecursively();
     attachRenderObject(newSlot);
     assert(_debugLifecycleState == _ElementLifecycle.active);
   }
 
-  static void _activateRecursively(Element element) {
-    assert(element._debugLifecycleState == _ElementLifecycle.inactive);
-    element.activate();
-    assert(element._debugLifecycleState == _ElementLifecycle.active);
-    element.visitChildren(_activateRecursively);
+  void _activateRecursively([Set<Type> newlyIntroducedInheritedWidgets = null]) {
+    assert(_debugLifecycleState == _ElementLifecycle.inactive);
+    newlyIntroducedInheritedWidgets = activate(newlyIntroducedInheritedWidgets);
+    assert(_debugLifecycleState == _ElementLifecycle.active);
+    visitChildren((Element child) {
+      child._activateRecursively(newlyIntroducedInheritedWidgets);
+    });
   }
 
-  void activate() {
+  /// Called when a previously de-activated widget (see [deactivate]) is reused
+  /// instead of being unmounted (see [unmount]).
+  ///
+  /// The argument is used to update the inheritance tree. It consists of the
+  /// Widget type of any newly introduced Inherited widgets that are now
+  /// available and that previously were not. It is null if there are no new
+  /// widget types.
+  ///
+  /// The return value is the argument to give to any children. Thus, for
+  /// instanced, the [InheritedElement] implementation of this method returns a
+  /// new set if it introduces a new widget. It can be null, meaning there are
+  /// no new widget types.
+  Set<Type> activate(Set<Type> newlyIntroducedInheritedWidgets) {
     assert(_debugLifecycleState == _ElementLifecycle.inactive);
     assert(widget != null);
     assert(depth != null);
     assert(!_active);
     _active = true;
     _updateInheritance();
+    if (newlyIntroducedInheritedWidgets != null) {
+      for (Type newInheritedWidget in newlyIntroducedInheritedWidgets) {
+        if (_unsatisfiedDependencies != null && _unsatisfiedDependencies.contains(newInheritedWidget))
+          dependenciesChanged(newInheritedWidget);
+      }
+    }
     assert(() { _debugLifecycleState = _ElementLifecycle.active; return true; });
+    return newlyIntroducedInheritedWidgets;
   }
 
+  // TODO(ianh): Define activation/deactivation thoroughly (other methods point
+  // here for details).
   void deactivate() {
     assert(_debugLifecycleState == _ElementLifecycle.active);
     assert(widget != null);
     assert(depth != null);
     assert(_active);
-    if (_dependencies != null) {
+    if (_dependencies != null && _dependencies.length > 0) {
       for (InheritedElement dependency in _dependencies)
         dependency._dependents.remove(this);
       _dependencies.clear();
     }
+    _unsatisfiedDependencies?.clear();
```

Comment:
Here you clear out `_unsatisfiedDependencies`, which means it will always be empty on line 1146.